src/hardware/adlib.cpp

```diff
@@ -39,19 +40,24 @@
 #define OPL2_INTERNAL_FREQ    3600000   // The OPL2 operates at 3.6MHz
 #define OPL3_INTERNAL_FREQ    14400000  // The OPL3 operates at 14.4MHz
 
+static Adlib::Module *module = nullptr;
+static AdlibGold *adlib_gold = nullptr;
+
+constexpr auto render_frames = 1024;
+
 namespace OPL2 {
 	#include "opl.cpp"
 
 struct Handler : public Adlib::Handler {
 	virtual void WriteReg(uint32_t reg, uint8_t val) { adlib_write(reg, val); }
 	virtual uint32_t WriteAddr(io_port_t, uint8_t val) { return val; }
 
-	virtual void Generate(mixer_channel_t &chan, const uint16_t samples)
+	virtual void Generate(mixer_channel_t &chan, const uint16_t frames)
 	{
-		int16_t buf[1024];
-		int remaining = samples;
+		int16_t buf[render_frames];
```

Comment:
Most of these Generate() calls will be asking for ~48 frames, but this 1024-item buffer `int16_t buf[render_frames];` gets created and destroyed every millsecond on the stack. I wonder if we should/could hoist it out into a static or make it a handler member variable? (maybe a std::array, similar to other refactor audio code?

Not asking for change, for your assessment.

Reply:
Could be done, but it's not dynamically allocated, so there's no performance penalty involved, unless I'm missing something? Technically the "allocation" happens by updating the stack pointer with a different value, no? I guess I like to keep them inside the function close to where they're used.

Reply:
> Technically the "allocation" happens by updating the stack pointer with a different value, no? I guess I like to keep them inside the function close to where they're used.

Thanks, @johnnovak - right on. It's also a perfect case for deliberately *not* zero or default-init'ing the array (just as a side comment, realize you're not init'ing here!).